File: 1/total_distance.ts
import * as fs from 'fs';

function parseInput(filename: string): { leftList: number[], rightList: number[] } {
    const fileContent = fs.readFileSync(filename, 'utf-8');
    const lines = fileContent.trim().split('\n');
    
    const leftList: number[] = [];
    const rightList: number[] = [];
    
    lines.forEach(line => {
        const [left, right] = line.trim().split(/\s+/).map(Number);
        leftList.push(left);
        rightList.push(right);
    });
    
    return { leftList, rightList };
}

function totalDistance(leftList: number[], rightList: number[]): number {
    return [...leftList]
        .sort((a, b) => a - b)
        .map((num, i) => Math.abs(num - [...rightList].sort((a, b) => a - b)[i]))
        .reduce((sum, curr) => sum + curr, 0);
}

function similarityScore(leftList: number[], rightList: number[]): number {
    return leftList.map((num, i) => num * rightList.filter(n => n === num).length).reduce((sum, curr) => sum + curr, 0);
}

const { leftList, rightList } = parseInput('1-input.txt');
console.log(totalDistance(leftList, rightList));
console.log(similarityScore(leftList, rightList));

File: 1/total_distance.rb
# 1. Parse input
# 2. Find pairs (lowest with lowest and so on)
# 3. Calculate the distance between the pairs
# 4. Find total distance between left list and right list (sum of distances)

def parse_input(filename)
  left_list = []
  right_list = []

  File.open(filename).each_line do |line|
    left, right = line.split(" ").map(&:to_i)
    left_list << left
    right_list << right
  end

  [left_list, right_list]
end

def total_distance(filename)
  left_list, right_list = parse_input(filename)
  left_list.sort!
  right_list.sort!

  distances = []
  left_list.each_with_index do |left, index|
    distances << (left - right_list[index]).abs
  end

  distances.sum
end

puts total_distance("1-input.txt")
